In [2]:
import pandas as pd

In [4]:

# Especifica la ruta y nombre del archivo CSV
archivo_csv = "combined_earthquake_data.csv"

# Lee el archivo CSV con pandas
df = pd.read_csv(archivo_csv)

# Muestra el contenido del DataFrame
df.head()


dt_sismo fecha_sismo hora_sismo  latitude  longitude  \
0  2012-01-01 18:50:43+00:00  2012-01-01   18:50:43   -26.913    -71.105   
1  2012-01-01 21:58:16+00:00  2012-01-01   21:58:16   -34.920    -72.759   
2  2012-01-03 21:50:22+00:00  2012-01-03   21:50:22   -36.994    -74.084   
3  2012-01-05 00:54:31+00:00  2012-01-05   00:54:31   -45.992    -76.472   
4  2012-01-05 09:17:35+00:00  2012-01-05   09:17:35   -27.921    -71.143   

   profundidad  mag                            ubicacion   pais peligrosidad  
0         45.2  4.6          91 km WNW of Copiapó, Chile  Chile     moderado  
1         34.7  4.1      55 km NW of Constitución, Chile  Chile     moderado  
2         25.1  4.6           73 km WNW of Arauco, Chile  Chile     moderado  
3         10.0  5.3  290 km W of Puerto Chacabuco, Chile  Chile       fuerte  
4         45.8  4.6         81 km NNW of Vallenar, Chile  Chile     moderado

In [5]:
# Obtiene los distintos tipos de valores y sus conteos en la columna "peligrosidad"
conteo_valores = df["peligrosidad"].value_counts()
conteo_valores

peligrosidad
moderado      24535
leve           2725
fuerte         1128
muy fuerte       10
Name: count, dtype: int64

In [11]:
from sklearn.model_selection import train_test_split

# Dividir el conjunto de datos en características (X) y variable objetivo (y)
X = df[['mag', 'profundidad']]  # Características
y = df["peligrosidad"]  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba (proporción 80:20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [12]:
from sklearn.ensemble import RandomForestClassifier

# Crear una instancia del modelo de Bosque Aleatorio
rf_model = RandomForestClassifier(random_state=42)

# Entrenar el modelo utilizando el conjunto de entrenamiento
rf_model.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

In [15]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

# Realizar predicciones en el conjunto de prueba
y_pred = rf_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión:", accuracy)

# Calcular el recall del modelo
recall = recall_score(y_test, y_pred, average="macro")
print("Recall:", recall)

from sklearn.metrics import classification_report
# Evaluar el rendimiento del modelo
print(classification_report(y_test, y_pred))


Precisión: 1.0
Recall: 1.0
              precision    recall  f1-score   support

      fuerte       1.00      1.00      1.00       208
        leve       1.00      1.00      1.00       549
    moderado       1.00      1.00      1.00      4920
  muy fuerte       1.00      1.00      1.00         3

    accuracy                           1.00      5680
   macro avg       1.00      1.00      1.00      5680
weighted avg       1.00      1.00      1.00      5680



In [16]:
from sklearn.model_selection import train_test_split

# Dividir el conjunto de datos en características (X) y variable objetivo (y)
X = df[['mag', 'profundidad']]  # Características
y = df["peligrosidad"]  # Variable objetivo

# Dividir los datos en conjuntos de entrenamiento y prueba (proporción 80:20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Dividir el conjunto de entrenamiento en conjuntos de entrenamiento y validación (proporción 75:25)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.25, random_state=42)

# Imprimir las formas de los conjuntos de datos resultantes
print("Forma de X_train:", X_train.shape)
print("Forma de y_train:", y_train.shape)
print("Forma de X_val:", X_val.shape)
print("Forma de y_val:", y_val.shape)
print("Forma de X_test:", X_test.shape)
print("Forma de y_test:", y_test.shape)


Forma de X_train: (17038, 2)
Forma de y_train: (17038,)
Forma de X_val: (5680, 2)
Forma de y_val: (5680,)
Forma de X_test: (5680, 2)
Forma de y_test: (5680,)


In [17]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, recall_score, classification_report

# Crear una instancia del modelo de Bosque Aleatorio
rf_model = RandomForestClassifier(random_state=42)

# Entrenar el modelo utilizando el conjunto de entrenamiento
rf_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = rf_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión:", accuracy)

# Calcular el recall del modelo
recall = recall_score(y_test, y_pred, average="macro")
print("Recall:", recall)

# Evaluar el rendimiento del modelo utilizando el informe de clasificación
classification_report = classification_report(y_test, y_pred)
print("Informe de clasificación:\n", classification_report)


Precisión: 1.0
Recall: 1.0
Informe de clasificación:
               precision    recall  f1-score   support

      fuerte       1.00      1.00      1.00       208
        leve       1.00      1.00      1.00       549
    moderado       1.00      1.00      1.00      4920
  muy fuerte       1.00      1.00      1.00         3

    accuracy                           1.00      5680
   macro avg       1.00      1.00      1.00      5680
weighted avg       1.00      1.00      1.00      5680



Validación cruzada utilizando k-fold cross-validation

In [18]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

# Crear una instancia del modelo de Bosque Aleatorio
rf_model = RandomForestClassifier(random_state=42)

# Aplicar k-fold cross-validation con 5 folds
scores = cross_val_score(rf_model, X_train, y_train, cv=5) # Asegúrate de ajustar el número de folds (cv) según tus necesidades. Cuanto mayor sea el número de folds, más robusta será la evaluación, pero también aumentará el tiempo de ejecución.

# Imprimir los resultados de validación cruzada
print("Resultados de validación cruzada:")
print("Precisión promedio:", scores.mean())
print("Desviación estándar de precisión:", scores.std())


Resultados de validación cruzada:
Precisión promedio: 0.9996478184240951
Desviación estándar de precisión: 0.0002875550594707632


El resultado de la validación cruzada muestra una precisión promedio de aproximadamente 0.9996 y una desviación estándar de precisión de aproximadamente 0.0003. Esto indica que el modelo de Bosque Aleatorio tiene un rendimiento consistente y preciso en diferentes divisiones de los datos de entrenamiento.

Código para SVM (Support Vector Machine)

In [ ]:
# import pandas as pd
# from sklearn.svm import SVC
# from sklearn.model_selection import train_test_split
# from sklearn.metrics import classification_report
# from sklearn.preprocessing import StandardScaler

# # Cargar el dataset
# df = pd.read_csv("combined_earthquake_data.csv")

# # Seleccionar las características (magnitud y profundidad) y la variable objetivo (peligrosidad)
# X = df[['mag', 'profundidad']]
# y = df['peligrosidad']

# # Normalizar las características
# scaler = StandardScaler()
# X = scaler.fit_transform(X)

# # Dividir el conjunto de datos en conjuntos de entrenamiento y prueba
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# # Crear y entrenar el modelo de SVM
# model_svm = SVC(random_state=42)
# model_svm.fit(X_train, y_train)

# # Realizar predicciones en el conjunto de prueba
# y_pred = model_svm.predict(X_test)

# # Evaluar el rendimiento del modelo
# print(classification_report(y_test, y_pred))
